msvc: Fix module freezing

stinos opened the pull request:
Brought up in #10548, freezing of 'real' modules didn't actually work for a couple of reasons; probably implementation was only tested once on a simple file.
Note:
- makemanifest.py relies on qstrdefs.preprocessed.h containing all possible qstrs. This happens to be how py.mk creates qstrdefs.preprocessed.h, but for the msbuild build there were 2 separate files which I now concatenated. Alternative would be to alter makemanifest and give it arguments to pass multiple files but that seems overkill?
- main issue was backslash vs forward slash in paths. In mpy-tool I fixed the occurrences which made building of the dev variant (i.e. freezing of uasyncio) fail and some others I noticed, but I'm not quite sure that's really all of it. Alternative would be to have mpy-cross generate .mpy files with forward slashes only and then everything else probably just works? Would that be preferrable?

stinos committed:
windows/msvc: Remove variant suffix from executable filename.

This is in line with the change made for other ports in d53c3b6a: since
the default output directory already includes the variant name in it
there's no need to add it to the executable as well.

stinos committed:
windows/msvc: Make mpy-cross independent of micropython variant.

The variant.props may have incompatible build options which break
the mpy-cross build and in any case mpy-cross has nothing to do
with variant support.

stinos committed:
windows/msvc: Fix qstr generation dependency.

The mpversion.h file must exist before py/ source can be preprocessed,
but this went unnoticed because micropython.vcxproj always calls
MakeVersionHdr before MakeQstrDefs.

stinos committed:
windows/msvc: Fix module freezing.

Make this more generally useful and in line with what the mingw
and unix ports do: 16bit dig size to work on 32bit ports, a
self-contained qstrdefs.preprocessed.h because makemanifest.py
uses that, and a dev variant which effectively puts this to use:
previously the uasyncio module wasn't frozen but instead tests
ran by importing it from the extmod/ directory.

dpgeorge commented:
> Alternative would be to alter makemanifest and give it arguments to pass multiple files but that seems overkill?

I think what you've done in this PR is good, to concat the output into one file, to pass to `makemanifest.py`.


> Alternative would be to have mpy-cross generate .mpy files with forward slashes only and then everything else probably just works? Would that be preferrable?

Yes I think that would be preferable, because then a .mpy file generated on Windows will still work on a *nix based system that uses '/'.  (I'm assuming here that at the moment running `mpy-cross` under Windows to build, eg, uasyncio will encode, eg, `uasyncio\event.py` in the .mpy file, whereas it would be preferable to have that `uasyncio/event.py`.)

stinos commented:
> I'm assuming here that at the moment running mpy-cross under Windows to build, eg, uasyncio will encode, eg, uasyncio\event.py in the .mpy file

Yes that's exactly the issue. Agreed that forcing mpy-cross to use `/` is the way to go, such that no matter where a .mpy comes from it has (roughly) the same content.

stinos committed:
mpy-cross: Force forward slashes in paths.

Code in tools/mpy-tool.py and py/frozenmod.c relies on the source file
path encoded in a .mpy file to have forward slashes (e.g. by searching
for '/__init__.py').  Enforce that when creating these files, thereby
fixing import of .mpy files and frozen modules not working before
because they could have backslashes when built with the windows port.

dpgeorge commented on mpy-cross/main.c:
```diff
@@ -242,7 +251,7 @@ MP_NOINLINE int main_(int argc, char **argv) {
                 }
                 a += 1;
-                source_file = argv[a];
+                source_file = backslash_to_forwardslash(argv[a]);
```
I was hoping this bit wasn't needed, because it would leave an "escape hatch" where someone could, if they really needed, force the use of backslash in the path name.

But I guess you needed to make this change because in some places in the build it would pass `-s <path-with-backslash>` here, and that needs to be converted.

stinos replied:
Indeed `source_file` is exactly is the one which needs to be fixed when using make_manifest.py: the argument to mpy-cross looks like `C:\path\to\uasyncio/__init__.py` or similar. Don't think there's a better way?

dpgeorge replied:
What we're effectively doing now is making a new restriction/definition of .mpy files:
- the source name stored in a .mpy file must use `/` to separate bits of the path

And we need to do this because a lot of the tools (eg `mpy-tool.py` and the way it freezes .mpy files) have implicitly made this assumption.  So now we will enforce it via `mpy-cross`.

I guess that's acceptable then.

The alternative is that the source name in .mpy files can contain `/` and `\` and all tools must handle both characters (what you were trying to fix before).  But that makes the tools complicated.  So I think it's fair enough to make this restriction and use only `/`.

stinos replied:
> What we're effectively doing now is making a new restriction/definition of .mpy files:

Indeed, but that restriction isn't really 'new' because all tools use it already

dpgeorge commented:
Thanks for this, a good set of fixes!
